mesosphere/charts
# Add kommander-karma chart

https://jira.mesosphere.com/browse/DCOS-59076

This PR depends on the `mtls-proxy` chart from #195. It also depends on the karma chart from #216.

This is part of a project to add support to Kommander for centralized monitoring of Konvoy clusters.

Konvoy's Prometheus addon provides an Alertmanager that handles alerts on that cluster. This PR adds a new chart that deploys [Karma](https://github.com/prymitive/karma), which provides a centralized alert dashboard for multiple Alertmanagers, and declares a federated mTLS proxy for the Alertmanager on each managed Konvoy. Karma consumes each Alertmanager's API through its respective proxy, and provides an alert dashboard that reflects the state of all managed Konvoys' Alertmanagers.

This does not include a controller for service discovery, which will be added later. In the meantime, clusters can be added manually by editing the Karma configmap and adding an entry to `alertmanagers` with the proxy's external address and TLS config. See the testing section below for an example.

This PR adds the `kommander-karma` chart, which deploys the Karma Helm chart, a `mtls-proxy` `FederatedAddon` that's configured to proxy Alertmanager, and the certificate secrets necessary to secure their communication. The `kommander-karma` chart will be added as a subchart of `kommander` in a later PR.

## Testing

This is how I tested the PR:

1. Deploy a Konvoy cluster.
1. Join the cluster to the kubefed federation.
    * `kubefedctl join kommander --kubefed-namespace=kommander --host-cluster-context=kommander-admin@kommander --host-cluster-name=kommander --cluster-context=kommander-admin@kommander`
1. Deploy the kommander-karma helm chart.
    * `helm install stable/kommander-karma --name kommander-kommander-karma --namespace kommander --set alertmanagerAddress="prometheus-kubeaddons-prom-alertmanager.kubeaddons.svc.cluster.local:9093"`
1. Retrieve the external hostname for the proxy's load balancer.
    * `kubectl -n kommander-system get service kommander-kommander-karma-proxy-kubeaddons-mtls-proxy -o jsonpath="{ .status.loadBalancer.ingress[0].hostname}"`
1. Add the proxy to Karma's configuration as an Alertmanager.
    1. `kubectl -n kommander edit configmap kommander-kommander-karma-config`
    1. Remove the placeholder in `karma.conf`'s `alertmanager.servers` and replace it with this entry:
    ```yaml
      - name: kommander
        uri: https://<proxy-hostname>:443  # Replace with the actual proxy hostname from the previous step.
        tls:
          ca: /etc/certs/kommander-karma-client-tls/ca.crt
          cert: /etc/certs/kommander-karma-client-tls/tls.crt
          key: /etc/certs/kommander-karma-client-tls/tls.key
          insecureSkipVerify: true
    ```
1. Restart the karma deployment to reload its configuration.
    1. `kubectl scale -n kommander deployment kommander-kommander-karma --replicas=0`
    1. `kubectl scale -n kommander deployment kommander-kommander-karma --replicas=1`
1. Browse to `/kommander/monitoring/karma` and observe that a watchdog alert from the connected Alertmanager is present.
